ext/opcache/jit/zend_jit_trace.c

```diff
@@ -2293,35 +2326,62 @@ static zend_ssa *zend_jit_trace_build_tssa(zend_jit_trace_rec *trace_buffer, uin
 		ssa = &jit_extension->func_info.ssa;
 
 		while (phi) {
-			uint32_t t  = ssa_var_info[phi->ssa_var].type;
-			uint32_t t0 = ssa_var_info[phi->sources[0]].type;
-			uint32_t t1 = ssa_var_info[phi->sources[1]].type;
+			uint32_t t = ssa_var_info[phi->ssa_var].type;
 
-			if (t & MAY_BE_GUARD) {
-				if (tssa->vars[phi->ssa_var].alias != NO_ALIAS) {
-					/* pass */
-				} else if (((t0 | t1) & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF)) == (t & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF))) {
+			if ((t & MAY_BE_GUARD) && tssa->vars[phi->ssa_var].alias == NO_ALIAS) {
+				uint32_t t0 = ssa_var_info[phi->sources[0]].type;
+				uint32_t t1 = ssa_var_info[phi->sources[1]].type;
+
+				if (((t0 | t1) & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF)) == (t & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF))) {
 					if (!((t0 | t1) & MAY_BE_GUARD)) {
 						ssa_var_info[phi->ssa_var].type = t & ~MAY_BE_GUARD;
 					}
 				} else if ((t1 & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF)) == (t & (MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF))) {
 					if (!(t1 & MAY_BE_GUARD)
 					 || is_checked_guard(tssa, ssa_opcodes, phi->sources[1], phi->ssa_var)) {
 						ssa_var_info[phi->ssa_var].type = t & ~MAY_BE_GUARD;
-						ssa_var_info[phi->sources[0]].type = t | MAY_BE_GUARD;
+						ssa_var_info[phi->sources[0]].type =
+							(t0 & MAY_BE_ARRAY_KEY_ANY) |
+							(t & ~(MAY_BE_ARRAY_KEY_ANY|MAY_BE_PACKED_GUARD)) |
+							MAY_BE_GUARD;
```

Comment:
Rather than handling `MAY_BE_ARRAY_KEY_ANY` specially here, is it possible to base this on the `MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF` part of the mask that is being compared/guarded?
```
uint32_t type_mask = MAY_BE_ANY|MAY_BE_UNDEF|MAY_BE_REF;
ssa_var_info[phi->sources[0]].type = (t0 & ~type_mask) | (t & type_mask) | MAY_BE_GUARD;
```
Or do I misunderstand the purpose of this code?

Reply:
Agree.